File: frontend/web_angular/src/app/admin/analytics/services/analytics.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable, forkJoin, of } from 'rxjs';
import { map, switchMap } from 'rxjs/operators';
import { AnalyticsResponse, KpiData } from '../analytics.types';

@Injectable({
  providedIn: 'root'
})
export class AnalyticsService {
  private jobsApiUrl = 'http://localhost:8888/joboffer-service/api/analytics/kpi';
  private usersApiUrl = 'http://localhost:8888/auth-service/analytics/users';
  private applicationsApiUrl = 'http://localhost:8888/application-service/api/analytics/applications';
  private apiUrl = 'http://localhost:8888/joboffer-service/api/analytics';

  constructor(private http: HttpClient) {}

  /**
   * Get KPI data from backend
   */
  getKpiData(): Observable<AnalyticsResponse<KpiData>> {
    return forkJoin({
      jobs: this.http.get<any>(this.jobsApiUrl),
      users: this.http.get<any>(this.usersApiUrl),
      applications: this.http.get<any>(this.applicationsApiUrl)
    }).pipe(
      map(result => ({
        data: {
          totalJobs: result.jobs.totalJobs || 0,
          jobsLast7Days: result.jobs.jobsLast7Days || 0,
          totalUsers: result.users.totalUsers || 0,
          newUsersLast7Days: result.users.newUsersLast7Days || 0,
          totalApplications: result.applications.totalApplications || 0,
          avgApplicationsPerJob: result.applications.avgApplicationsPerJob || 0
        },
        timestamp: new Date().toISOString()
      }))
    );
  }

  /**
   * Get jobs posted over time
   */
  getJobsOverTime(startDate?: string, endDate?: string): Observable<any> {
    let url = `${this.apiUrl}/jobs-over-time`;
    const params: any = {};
    
    if (startDate) params.startDate = startDate;
    if (endDate) params.endDate = endDate;
    
    return this.http.get<any>(url, { params }).pipe(
      map(response => response.data || response)
    );
  }

  /**
   * Get application funnel data
   */
  getApplicationFunnel(): Observable<any> {
    return this.http.get<any>('http://localhost:8888/application-service/api/analytics/funnel').pipe(
      map(response => ({
        data: {
          applications: response.applications || 0,
          interviews: response.interviews || 0,
          hires: response.hires || 0
        }
      }))
    );
  }

  /**
   * Get applications per job data
   */
  getAppsPerJob(): Observable<any> {
    return forkJoin({
      appsData: this.http.get<any>('http://localhost:8888/application-service/api/analytics/apps-per-job'),
      allJobs: this.http.get<any>('http://localhost:8888/joboffer-service/api/jobs')
    }).pipe(
      map(result => {
        // Create a map of jobId -> jobTitle
        const jobMap = new Map<number, string>();
        if (result.allJobs && Array.isArray(result.allJobs)) {
          result.allJobs.forEach((job: any) => {
            jobMap.set(job.id, job.title);
          });
        }

        // Enrich appsData with job titles
        const enrichedData = result.appsData.data.map((item: any) => ({
          jobId: item.jobId.toString(),
          jobTitle: jobMap.get(item.jobId) || `Job #${item.jobId}`,
          count: item.count
        }));

        return {
          data: enrichedData,
          timestamp: new Date().toISOString()
        };
      })
    );
  }

  /**
   * Get top companies by activity
   */
  getTopCompanies(): Observable<any> {
    // First, get company data from joboffer-service
    return this.http.get<any>('http://localhost:8888/joboffer-service/api/analytics/top-companies').pipe(
      switchMap(response => {
        const companies = response.data || [];
        
        if (companies.length === 0) {
          return of({
            data: [],
            timestamp: new Date().toISOString()
          });
        }
        
        // For each company, we need to count total applications
        const enrichmentRequests = companies.map((company: any) => {
          // Get application counts for all jobs from this company
          return this.http.post<any>(
            'http://localhost:8888/application-service/api/analytics/count-by-jobs',
            { jobIds: company.jobIds }
          ).pipe(
            map(appResponse => {
              const counts = appResponse.counts || {};
              const totalApplications = Object.values(counts).reduce((sum: number, count: any) => sum + count, 0);
              const avgAppsPerJob = company.jobCount > 0 ? totalApplications / company.jobCount : 0;
              
              return {
                company: company.company,
                jobCount: company.jobCount,
                totalApplications: totalApplications,
                avgAppsPerJob: Math.round(avgAppsPerJob * 10) / 10 // Round to 1 decimal
              };
            })
          );
        });

        // Execute all requests in parallel and return combined data
        return forkJoin(enrichmentRequests).pipe(
          map(enrichedCompanies => ({
            data: enrichedCompanies,
            timestamp: new Date().toISOString()
          }))
        );
      })
    );
  }

  /**
   * Get top job categories by job position
   */
  getTopCategories(): Observable<any> {
    // First, get category data from joboffer-service
    return this.http.get<any>('http://localhost:8888/joboffer-service/api/analytics/top-categories').pipe(
      switchMap(response => {
        const categories = response.data || [];
        
        if (categories.length === 0) {
          return of({
            data: [],
            timestamp: new Date().toISOString()
          });
        }
        
        // For each category, we need to count total applications
        const enrichmentRequests = categories.map((category: any) => {
          // Get application counts for all jobs in this category
          return this.http.post<any>(
            'http://localhost:8888/application-service/api/analytics/count-by-jobs',
            { jobIds: category.jobIds }
          ).pipe(
            map(appResponse => {
              const counts = appResponse.counts || {};
              const totalApplications = Object.values(counts).reduce((sum: number, count: any) => sum + count, 0);
              const ratio = category.jobCount > 0 ? totalApplications / category.jobCount : 0;
              
              return {
                category: category.category,
                jobCount: category.jobCount,
                totalApplications: totalApplications,
                ratio: Math.round(ratio * 10) / 10 // Round to 1 decimal
              };
            })
          );
        });

        // Execute all requests in parallel and return combined data
        return forkJoin(enrichmentRequests).pipe(
          map(enrichedCategories => ({
            data: enrichedCategories,
            timestamp: new Date().toISOString()
          }))
        );
      })
    );
  }

  /**
   * Get geographic distribution (jobs by location with application counts)
   */
  getGeographicDistribution(): Observable<any> {
    // First, get location data from joboffer-service
    return this.http.get<any>('http://localhost:8888/joboffer-service/api/analytics/geographic-distribution').pipe(
      switchMap(response => {
        const locations = response.data || [];
        
        if (locations.length === 0) {
          return of({
            data: [],
            timestamp: new Date().toISOString()
          });
        }
        
        // For each location, we need to count total applications
        const enrichmentRequests = locations.map((location: any) => {
          // Get application counts for all jobs in this location
          return this.http.post<any>(
            'http://localhost:8888/application-service/api/analytics/count-by-jobs',
            { jobIds: location.jobIds }
          ).pipe(
            map(appResponse => {
              const counts = appResponse.counts || {};
              const totalApplications = Object.values(counts).reduce((sum: number, count: any) => sum + count, 0);
              
              return {
                location: location.location,
                jobCount: location.jobCount,
                totalApplications: totalApplications
              };
            })
          );
        });

        // Execute all requests in parallel and return combined data
        return forkJoin(enrichmentRequests).pipe(
          map(enrichedLocations => ({
            data: enrichedLocations,
            timestamp: new Date().toISOString()
          }))
        );
      })
    );
  }

  /**
   * Get job posts alert data (current vs previous period)
   */
  getJobPostsAlert(days: number = 7): Observable<any> {
    return this.http.get<any>(`http://localhost:8888/joboffer-service/api/analytics/job-posts-alert?days=${days}`);
  }

  /**
   * Get applications alert data (current vs previous period)
   */
  getApplicationsAlert(days: number = 7): Observable<any> {
    return this.http.get<any>(`http://localhost:8888/application-service/api/analytics/applications-alert?days=${days}`);
  }

  /**
   * Get new users alert data (current vs previous period)
   */
  getNewUsersAlert(days: number = 7): Observable<any> {
    return this.http.get<any>(`http://localhost:8888/auth-service/analytics/users-alert?days=${days}`);
  }

  /**
   * Get interviews alert data (current vs previous period)
   */
  getInterviewsAlert(days: number = 7): Observable<any> {
    return this.http.get<any>(`http://localhost:8888/application-service/api/analytics/interviews-alert?days=${days}`);
  }

  /**
   * Get hiring rate alert data (current vs previous period)
   */
  getHiringAlert(days: number = 7): Observable<any> {
    return this.http.get<any>(`http://localhost:8888/application-service/api/analytics/hiring-alert?days=${days}`);
  }

  /**
   * Get all alert metrics in one call
   */
  getAllAlerts(days: number = 7): Observable<any> {
    return forkJoin({
      applications: this.getApplicationsAlert(days),
      jobPosts: this.getJobPostsAlert(days),
      newUsers: this.getNewUsersAlert(days),
      interviews: this.getInterviewsAlert(days),
      hiring: this.getHiringAlert(days)
    }).pipe(
      map(alerts => ({
        data: {
          totalApplications: {
            current: alerts.applications.current || 0,
            percentageChange: alerts.applications.percentageChange || 0,
            isAlert: alerts.applications.isAlert || false
          },
          jobPosts: {
            current: alerts.jobPosts.current || 0,
            percentageChange: alerts.jobPosts.percentageChange || 0,
            isAlert: alerts.jobPosts.isAlert || false
          },
          newUsers: {
            current: alerts.newUsers.current || 0,
            percentageChange: alerts.newUsers.percentageChange || 0,
            isAlert: alerts.newUsers.isAlert || false
          },
          interviews: {
            current: alerts.interviews.current || 0,
            percentageChange: alerts.interviews.percentageChange || 0,
            isAlert: alerts.interviews.isAlert || false
          },
          hiringRate: {
            current: alerts.hiring.currentRate || 0,
            percentageChange: alerts.hiring.percentageChange || 0,
            isAlert: alerts.hiring.isAlert || false
          }
        },
        timestamp: new Date().toISOString()
      }))
    );
  }
}

File: backend/joboffer-service/src/main/java/com/joboffer/jobofferservice/service/AnalyticsService.java
package com.joboffer.jobofferservice.service;

import com.joboffer.jobofferservice.model.JobOffer;
import com.joboffer.jobofferservice.repository.JobOfferRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.*;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Slf4j
public class AnalyticsService {

    private final JobOfferRepository jobOfferRepository;

    /**
     * Get KPI data: total jobs, jobs last 7 days, etc.
     */
    public Map<String, Object> getKpiData() {
        log.info("Calculating KPI data");

        // Total jobs (all time)
        long totalJobs = jobOfferRepository.count();

        // Jobs posted in last 7 days
        LocalDateTime sevenDaysAgo = LocalDateTime.now().minusDays(7);
        long jobsLast7Days = jobOfferRepository.countByCreatedAtAfter(sevenDaysAgo);

        Map<String, Object> kpiData = new HashMap<>();
        kpiData.put("totalJobs", totalJobs);
        kpiData.put("jobsLast7Days", jobsLast7Days);
        kpiData.put("totalUsers", 0); // Will be fetched from auth-service
        kpiData.put("newUsersLast7Days", 0); // Will be fetched from auth-service
        kpiData.put("totalApplications", 0); // Will be fetched from application-service
        kpiData.put("avgApplicationsPerJob", 0.0);

        log.info("KPI data calculated: totalJobs={}, jobsLast7Days={}", totalJobs, jobsLast7Days);

        return kpiData;
    }

    /**
     * Get jobs posted over time (date range)
     */
    public Map<String, Object> getJobsOverTime(String startDateStr, String endDateStr) {
        log.info("Getting jobs over time from {} to {}", startDateStr, endDateStr);

        LocalDate startDate = startDateStr != null ? 
            LocalDate.parse(startDateStr) : LocalDate.now().minusDays(30);
        LocalDate endDate = endDateStr != null ? 
            LocalDate.parse(endDateStr) : LocalDate.now();

        LocalDateTime startDateTime = startDate.atStartOfDay();
        LocalDateTime endDateTime = endDate.atTime(23, 59, 59);

        // Get all jobs in date range
        var jobs = jobOfferRepository.findByCreatedAtBetween(startDateTime, endDateTime);

        // Group by date and count
        Map<LocalDate, Long> jobsByDate = jobs.stream()
            .collect(Collectors.groupingBy(
                job -> job.getCreatedAt().toLocalDate(),
                Collectors.counting()
            ));

        // Convert to list format
        List<Map<String, Object>> data = jobsByDate.entrySet().stream()
            .map(entry -> {
                Map<String, Object> point = new HashMap<>();
                point.put("date", entry.getKey().toString());
                point.put("count", entry.getValue());
                return point;
            })
            .sorted(Comparator.comparing(m -> (String) m.get("date")))
            .collect(Collectors.toList());

        Map<String, Object> response = new HashMap<>();
        response.put("data", data);
        response.put("timestamp", LocalDateTime.now().toString());

        return response;
    }

    /**
     * Get top companies by job count and return recruiter IDs
     * Returns: company name, job count, recruiter ID
     */
    public Map<String, Object> getTopCompaniesByActivity() {
        log.info("Getting top companies by activity");

        // Get all jobs
        var allJobs = jobOfferRepository.findAll();

        // Group by company and collect recruiter IDs and job IDs
        Map<String, Map<String, Object>> companyData = new HashMap<>();
        
        for (var job : allJobs) {
            String company = job.getCompany();
            if (company == null || company.trim().isEmpty()) {
                company = "Unknown Company";
            }
            
            companyData.putIfAbsent(company, new HashMap<>());
            Map<String, Object> data = companyData.get(company);
            
            // Add recruiter ID (we'll use first one found for this company)
            if (!data.containsKey("recruiterId")) {
                data.put("recruiterId", job.getRecruiterId());
            }
            
            // Collect job IDs
            @SuppressWarnings("unchecked")
            List<Long> jobIds = (List<Long>) data.getOrDefault("jobIds", new ArrayList<>());
            jobIds.add(job.getId());
            data.put("jobIds", jobIds);
            
            // Count jobs
            data.put("jobCount", jobIds.size());
        }

        // Convert to list and sort by job count
        List<Map<String, Object>> companies = companyData.entrySet().stream()
            .map(entry -> {
                Map<String, Object> company = new HashMap<>();
                company.put("company", entry.getKey());
                company.put("jobCount", entry.getValue().get("jobCount"));
                company.put("recruiterId", entry.getValue().get("recruiterId"));
                company.put("jobIds", entry.getValue().get("jobIds"));
                return company;
            })
            .sorted((a, b) -> Integer.compare((Integer) b.get("jobCount"), (Integer) a.get("jobCount")))
            .limit(10)
            .collect(Collectors.toList());

        Map<String, Object> response = new HashMap<>();
        response.put("data", companies);

        log.info("Top companies calculated: {} companies", companies.size());
        return response;
    }

    /**
     * Get top job categories (grouped by job position)
     */
    public Map<String, Object> getTopJobCategories() {
        log.info("Calculating top job categories by job position");

        List<JobOffer> allJobs = jobOfferRepository.findAll();
        
        // Group jobs by jobPosition
        var jobsByPosition = allJobs.stream()
            .filter(job -> job.getJobPosition() != null && !job.getJobPosition().trim().isEmpty())
            .collect(java.util.stream.Collectors.groupingBy(JobOffer::getJobPosition));

        // Create category data with job counts and jobIds
        var categoryData = jobsByPosition.entrySet().stream()
            .map(entry -> {
                String category = entry.getKey();
                List<JobOffer> jobs = entry.getValue();
                
                // Collect all job IDs for this category
                List<Long> jobIds = jobs.stream()
                    .map(JobOffer::getId)
                    .collect(java.util.stream.Collectors.toList());
                
                Map<String, Object> categoryInfo = new HashMap<>();
                categoryInfo.put("category", category);
                categoryInfo.put("jobCount", jobs.size());
                categoryInfo.put("jobIds", jobIds);
                
                return categoryInfo;
            })
            .sorted((a, b) -> Integer.compare(
                (Integer) b.get("jobCount"), 
                (Integer) a.get("jobCount")
            ))
            .limit(10) // Top 10 categories
            .collect(java.util.stream.Collectors.toList());

        Map<String, Object> response = new HashMap<>();
        response.put("data", categoryData);

        log.info("Top job categories calculated: {} categories", categoryData.size());
        return response;
    }

    /**
     * Get geographic distribution (jobs grouped by location with application counts)
     */
    public Map<String, Object> getGeographicDistribution() {
        log.info("Calculating geographic distribution by location");

        List<JobOffer> allJobs = jobOfferRepository.findAll();
        
        // Group jobs by location
        var jobsByLocation = allJobs.stream()
            .filter(job -> job.getLocation() != null && !job.getLocation().trim().isEmpty())
            .collect(java.util.stream.Collectors.groupingBy(JobOffer::getLocation));

        // Create location data with job counts and jobIds
        var locationData = jobsByLocation.entrySet().stream()
            .map(entry -> {
                String location = entry.getKey();
                List<JobOffer> jobs = entry.getValue();
                
                // Collect all job IDs for this location
                List<Long> jobIds = jobs.stream()
                    .map(JobOffer::getId)
                    .collect(java.util.stream.Collectors.toList());
                
                Map<String, Object> locationInfo = new HashMap<>();
                locationInfo.put("location", location);
                locationInfo.put("jobCount", jobs.size());
                locationInfo.put("jobIds", jobIds);
                
                return locationInfo;
            })
            .sorted((a, b) -> Integer.compare(
                (Integer) b.get("jobCount"), 
                (Integer) a.get("jobCount")
            ))
            .limit(10) // Top 10 locations
            .collect(java.util.stream.Collectors.toList());

        Map<String, Object> response = new HashMap<>();
        response.put("data", locationData);

        log.info("Geographic distribution calculated: {} locations", locationData.size());
        return response;
    }

    /**
     * Get jobs count for a specific date range
     */
    private long getJobsCountForPeriod(LocalDateTime startDate, LocalDateTime endDate) {
        return jobOfferRepository.findAll().stream()
            .filter(job -> job.getCreatedAt() != null)
            .filter(job -> !job.getCreatedAt().isBefore(startDate) && !job.getCreatedAt().isAfter(endDate))
            .count();
    }

    /**
     * Calculate percentage change
     */
    private double calculatePercentageChange(long current, long previous) {
        if (previous == 0) {
            return current > 0 ? 100.0 : 0.0;
        }
        return ((double) (current - previous) / previous) * 100;
    }

    /**
     * Get job posts alert (current period vs previous period)
     */
    public Map<String, Object> getJobPostsAlert(int days) {
        log.info("Calculating job posts alert for {} days", days);

        LocalDateTime now = LocalDateTime.now();
        LocalDateTime currentPeriodStart = now.minusDays(days);
        LocalDateTime previousPeriodStart = now.minusDays(days * 2);
        LocalDateTime previousPeriodEnd = currentPeriodStart;

        long currentPeriodCount = getJobsCountForPeriod(currentPeriodStart, now);
        long previousPeriodCount = getJobsCountForPeriod(previousPeriodStart, previousPeriodEnd);
        
        double percentageChange = calculatePercentageChange(currentPeriodCount, previousPeriodCount);
        
        Map<String, Object> response = new HashMap<>();
        response.put("current", currentPeriodCount);
        response.put("previous", previousPeriodCount);
        response.put("percentageChange", Math.round(percentageChange * 10.0) / 10.0);
        response.put("isAlert", percentageChange < -5.0); // Alert if decrease > 5%

        log.info("Job posts alert: current={}, previous={}, change={}%", 
                currentPeriodCount, previousPeriodCount, percentageChange);

        return response;
    }
}
